### ISSUE TITLE
Performance issue with Hololens

### ISSUE THREAD

**Issue by rishi-rranjan** [3DStreamingToolkit/3DStreamingToolkit#296]
## Report

- [ ] I have searched [existing issues](https://github.com/3DStreamingToolkit/3DStreamingToolkit/issues) and this is not a duplicate.

## Environment

- OS Version: Windows 10 Pro

- SDK Version: 10.0.17134.0 

- Visual Studio Version: 2017

## Issues and Steps to Reproduce

- Launch  Release 2.0 client on Hololens
- Launch Multithreaded server on PC

## Expected Behavior
- Delay should be minimal since server is OnPrem. 

- Minimal delay seen when Client is compiled with SHOW_DEBUG_INFO  is 250 ms. It goes up to 350 to 500 ms. Hololens experience is extremely bad. 

## Link to Code
***If you have some code that maintainers can clone/test for themselves, bugs can be resolved much faster. Please paste a link here.***

**Comment by rishi-rranjan** [comment]
To start debugging the issue, compiled webrtc-extensions-3dstk with SDK 10.0.17134.0.  Changed webrtc/windows/org/webrtc/wrapper/MediaSourceHelper.cc to print the debug prints to log the _frames size. 

But after compiling the StreamingDirectXHololensClient from Release 2.0 with above libraries, streaming is not working. Logs show lot of errors so need to look these to debug further. Attaching the log. Will look at it later today. 
 
[logs.txt](https://github.com/3DStreamingToolkit/3DStreamingToolkit/files/2415433/logs.txt)

**Comment by rishi-rranjan** [comment]
The error just seems like a random issue when launching the application from VS 2017 for debugging. I need the logs from MediaSourceHelper.cc so I want to keep using VS 2017. 

@phongcao as you mentioned, the frame size is 0 as sample duration is set as per frame size. There are some instances where Sample duration is high but it settles down to around 50. I see that most of the time FPS is 20 and Latency is > 300 ms.  

Plan to take this as follows: 

1. Print FPS and latency in the debug log with sample duration. 
2. Debug matching of prediction time with Holographic frame to make sure it's working as expected. 
3. If above 2 don't give any clue, enable RTC logs and start tracing packets at each API call in WebRTC. 

![sampleduration](https://user-images.githubusercontent.com/5673695/46031356-83eca780-c0ad-11e8-9241-f341418f25c8.png)
![frames_size](https://user-images.githubusercontent.com/5673695/46031482-ca420680-c0ad-11e8-8292-21228f1e5aa9.png)

**Comment by phongcao** [comment]
Your PC might have some issue with the video encoder since our Multithreaded sample runs at 40 - 50 fps normally. Can you reach 60 fps using the DirectX Win32 client?

**Comment by rishi-rranjan** [comment]
No even with DirectX Win32, I am only getting 20 FPS. Will debug it from PC side also.

**Comment by rishi-rranjan** [comment]
The PC keeps getting Intel display adapter even if we uninstall it. Moved to a different PC and confirmed 60 FPS on native windows client. Will try it with Hololens.

**Comment by rishi-rranjan** [comment]
Hololens looks much stable but it still has visible shaking. I will continue debugging this.

**Comment by phongcao** [comment]
Make sure you don't have the managed debugger attached to it while running since it might cause the stream to restart after a few minutes.

**Comment by rishi-rranjan** [comment]
@phongcao Thanks for the suggestion. I have seen this behavior. I am planning to enable RTC Stats to trace the end to end delay. I guess I will have to route it through ETW tracking. 

Currently with SHOW_DEBUG_INFO enabled, I get 50+ FPS and 120 ms to 150 ms latency. 

I expect following total latency in On prem deployments: 
1. I am ignoring time we get the input to the time WebRTC places it on wireless channel. 
2. Input transmission time from device to server(packets are less than 500 bytes which takes neglegible time to transmit): 1 ms 
3. Rendering time assuming 60 FPS on server( it should be 16.6 ms but since Unity is running in cycle, input packets too close to frame cycle will not be picked so I just double this for now): 33 ms
4. Encoding time: 5 ms
5. Transmission time over wireless(For full 2K, 18 Mbps is required. I see lower data transmitted from Server in 3Dstreamingtoolkit. For my purpose, I am taking 18 mbps as higher bound. This with 60 FPS takes 3 ms to transmit over the air assuming 100 mbps wifi link): 3 ms
6. Packet decode time: 5 ms
7. I am ignoring the time the client application runs to pick up the decoded packet for rendering. This will add to some delay. 

So I expect total end to end latency to be = 1 + 5 + 33 + 3 + 5 = 45 ms. 

Will validate and adjust these where extra latency is coming from and readjust it and come up with steps to get it closer to expected value. 

Will most probably pick this up next week.

**Comment by rishi-rranjan** [comment]
@phongcao Is there an option to print FPS and Latency in Unity client?

**Comment by phongcao** [comment]
No, there isn't. Another thing is that Unity client doesn't have image stabilization since there is no available API to grab the holographic frame.

**Comment by rishi-rranjan** [comment]
Update for  now: 
1. Was able to get the RTC Trace working on server side. 
2. rtc_tool is not compiling as part of the build so looking into it so I can use it to analyze the data.

**Comment by Holo-Krzysztof** [comment]
We're having performance problems as well; what is the easiest way to debug networking issues like packet loss?
